### PR TITLE
Proxy server to serve up varz on various ports

### PR DESCRIPTION
**Pull request by anurag6**

### CODE REVIEW

File: proto/forch_configuration.proto [hunk 1]
```diff
@@ -165,6 +165,15 @@ message EventClientConfig {
  * Varz interface section in Forch configuration
  */
 message VarzInterfaceConfig {
-  // port number to curl for varz list
-  int32 varz_port = 1;
+  // port number to curl for forch varz list
+  int32 forch_varz_port = 1;
```
Comment (grafnu) [inline proto/forch_configuration.proto:169]: this should be a generic map configuraiton, not specific to forch/faucet/gauge.  E.e., the config should have something like:

proxy_requests:
  proxy_port: 8032
  targets:
    faucet:
      port: 9032
    forch:
      port: 9036

File: forch/forch_metrics.py [hunk 2]
```diff
@@ -85,6 +107,31 @@ def get_metrics(self, path, params):
...
         return f"Error creating varz interface: {str(error)}"
+
+    def _get_url(self, server, port):
+        return 'http://' + str(server) + ':' + str(port)
+
+    def _register_metric_page(self, path, server, port):
+        self._metric_pages[path] = self._get_url(server, port)
+
+    def _register_metric_pages(self):
+        self._register_metric_page('faucet', LOCALHOST,
```
Comment (grafnu) [inline forch/forch_metrics.py:129]: this should all be dynamicallky pulled from config, not hard-coded

File: proto/forch_configuration.proto [hunk 3]
```diff
@@ -165,6 +165,15 @@ message EventClientConfig {
...
+
+  // port number to curl for faucet varz list
+  int32 faucet_varz_port = 2;
+
+  // port number to curl for gauge varz list
+  int32 gauge_varz_port = 3;
+
+  //port number for varz proxy
+  int32 proxy_varz_port = 4;
```
Comment (grafnu) [inline proto/forch_configuration.proto:178]: nothing about this proxy is specific to varz, so it shouldn't be encoded in the names

File: forch/forch_metrics.py [hunk 4]
```diff
@@ -32,6 +41,19 @@ def start(self):
         finally:
             threading.Thread(target=self._http_server.start_server(), daemon=True).start()
 
+        self._register_metric_pages()
+
+        self._proxy_server = forch.http_server.HttpServer(self._proxy_port)
+        try:
+            for page in self._metric_pages:
+                self._proxy_server.map_request(
+                    page, lambda path, params: self._get_path_metrics(path.split('/')[1]))
```
Comment (grafnu) [inline forch/forch_metrics.py:50]: what about the rest of the path?  Shouldn't this just be everything except the first bit?

Comment (anurag6) [reply forch/forch_metrics.py:50]: I didnt get it. Rest of the path?

Comment (grafnu) [reply forch/forch_metrics.py:50]: You take the input path, split it by /, and then only take [1] -- what about [2].. [n]?

File: forch/forch_metrics.py [hunk 5]
```diff
@@ -85,6 +107,31 @@ def get_metrics(self, path, params):
         """Return metric list in printable form"""
         return generate_latest(self._reg).decode('utf-8')
 
+    def _get_path_metrics(self, path):
```
Comment (grafnu) [inline forch/forch_metrics.py:110]: also, just realized you're perpetuating the wrong abstraction here.... this is just a web proxy, it has nothing to do with metrics... so likely should be pulled out into its own separate class.

File: forch/forch_metrics.py (hunk 2 above)
Comment (anurag6) [reply forch/forch_metrics.py]: Addressed.

File: proto/forch_configuration.proto (hunk 1 above)
Comment (anurag6) [reply proto/forch_configuration.proto]: Addressed.

File: proto/forch_configuration.proto (hunk 3 above)
Comment (anurag6) [reply proto/forch_configuration.proto]: Addressed.

File: forch/forch_metrics.py (hunk 4 above)
Comment (anurag6) [reply forch/forch_metrics.py]: Addressed.

File: forch/forch_metrics.py (hunk 5 above)
Comment (anurag6) [reply forch/forch_metrics.py]: Makes sense. Created new file.

**Comment by anurag6** [comment]
Made changes. PTAL @grafnu

File: bin/check_style [hunk 6]
```diff
@@ -12,7 +12,7 @@ if [ -z "$VIRTUAL_ENV" -a -d venv ]; then
     source venv/bin/activate
 fi
 
-PY_LIB=$VIRTUAL_ENV/lib/python3.7/site-packages/
+PY_LIB=$VIRTUAL_ENV/lib/python3.6/site-packages/
```
Comment (grafnu) [inline bin/check_style:15]: we've standardized on 3.7

File: forch/forch_proxy.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    """Class that implements the module that creates a proxy server"""
+
+    def __init__(self, proxy_config):
+        self._proxy_config = proxy_config
+        self._proxy_port = self._proxy_config.proxy_port or DEFAULT_PROXY_PORT
+        self._pages = {}
+        self._proxy_server = None
+
+    def start(self):
+        """Start serving varz"""
```
Comment (grafnu) [inline forch/forch_proxy.py:25]: this class should just be a generic web proxy, so there should be no mention of varz in this file...

File: forch/forchestrator.py [hunk 8]
```diff
@@ -128,6 +130,8 @@ def initialize(self):
         self._process_static_device_behavior()
         if self._faucetizer:
             self._faucetizer.flush_behavioral_config(force=True)
+        self._proxy = ForchProxy(self._config.proxy_server)
```
Comment (grafnu) [inline forch/forchestrator.py:133]: ForchProxy should only be instantiated/started if specifically enabled. So, you could either trigger it off of the port being defied and do away with DEFAULT_PROXY_PORT as a default, or trigger it off of the proxy config block being defined.  Either way, an empty config file should not instantiate the proxy at all.

File: proto/forch_configuration.proto [hunk 9]
```diff
@@ -165,6 +168,28 @@ message EventClientConfig {
  * Varz interface section in Forch configuration
  */
 message VarzInterfaceConfig {
-  // port number to curl for varz list
+  // port number for forch varz
   int32 varz_port = 1;
 }
+
+/*
+ * Target to scrape metrics from
```
Comment (grafnu) [inline proto/forch_configuration.proto:176]: it's a target for a port proxy, has nothing to do with metrics -- names/comments here and below should reflect that.

File: proto/forch_configuration.proto [hunk 10]
```diff
@@ -165,6 +168,28 @@ message EventClientConfig {
...
+ */
+message ProxyServerConfig {
+  // port number for varz proxy
+  int32 proxy_port = 1;
+
+  // list of targets to scrape metrics from
+  repeated MetricTarget targets = 2;
```
Comment (grafnu) [inline proto/forch_configuration.proto:183]: this should be a map of names that point at targets. It's nonsensical to have two names that are the same, and maps are strictly better than lists for sorting information unless you need to preserve ordering or allow duplicated.  See ProcessConfig for an example.  (Note that the value of the map should still be an object, even if it only has one item in it... as future-proofing)

File: topo/bond/forch.yaml [hunk 11]
```diff
@@ -9,6 +9,14 @@ site:
       port: 9020
 orchestration:
   static_device_placement: placement.yaml
+proxy_server:
```
Comment (grafnu) [inline topo/bond/forch.yaml:12]: also wouuld expect an update/change to the integration testing. It should be nearly transparent, just need to update the URL somewhere for what it's pointing at to test a raw varz access (which I believe is there somewhere).  Should be easy to check both proxied and raw access.

File: bin/check_style (hunk 6 above)
Comment (anurag6) [reply bin/check_style]: My mistake. Addressed.

File: forch/forch_proxy.py (hunk 7 above)
Comment (anurag6) [reply forch/forch_proxy.py]: Addressed.

File: proto/forch_configuration.proto (hunk 9 above)
Comment (anurag6) [reply proto/forch_configuration.proto]: Addressed.

File: proto/forch_configuration.proto (hunk 10 above)
Comment (anurag6) [reply proto/forch_configuration.proto]: Addressed.

File: topo/bond/forch.yaml (hunk 11 above)
Comment (anurag6) [reply topo/bond/forch.yaml:12]: Yup. Currently added to test_sol. Will replace other tests.

File: forch/forchestrator.py (hunk 8 above)
Comment (anurag6) [reply forch/forchestrator.py]: Addressed

**Comment by anurag6** [comment]
@grfnu PTAL

File: bin/stack_functions [hunk 12]
```diff
@@ -400,6 +400,11 @@ function validate_radius_varz {
 
 function test_varz_value {
     varz=$1
-    value=$( printf "%.0f" $(wget 0.0.0.0:8302 -O- | grep $varz | grep -v 'HELP\|TYPE' | awk '{print $2'}) )
+    if [ -z "$2" ]; then
+        path=8080/$2
```
Comment (grafnu) [inline bin/stack_functions:404]: not a path -- it's port + path ?

How does this possibly work?  -z tests for the value being zero?  I think you need a negative test somewhere.

File: bin/test_stack [hunk 13]
```diff
@@ -79,10 +79,10 @@ sudo ip link set t1sw2-eth28 up
...
 sleep 32
 test_forch -healthy
-test_varz_value process_state{process=\"faucet\"}
-test_varz_value process_state{process=\"gauge\"}
-test_varz_value process_state{process=\"forch\"}
-test_varz_value process_state{process=\"sleep\"}
+test_varz_value process_state{process=\"faucet\"} forch
+test_varz_value process_state{process=\"gauge\"} forch
+test_varz_value process_state{process=\"forch\"} forch
+test_varz_value process_state{process=\"sleep\"} forch
```
Comment (grafnu) [inline bin/test_stack:85]: make one of this native (empty), and then also add in a negative failing case (e.g. xyzzy)

File: proto/forch_configuration.proto [hunk 14]
```diff
@@ -173,23 +173,20 @@ message VarzInterfaceConfig {
...
+ * Proxy server config
  */
 message ProxyServerConfig {
-  // port number for varz proxy
+  // port number for proxy server
   int32 proxy_port = 1;
 
-  // list of targets to scrape metrics from
-  repeated MetricTarget targets = 2;
+  // map of targets
```
Comment (grafnu) [inline proto/forch_configuration.proto:182]: indicate what they map key is

File: proto/forch_configuration.proto [hunk 15]
```diff
@@ -179,7 +179,7 @@ message ProxyServerConfig {
   // port number for proxy server
   int32 proxy_port = 1;
 
-  // map of targets
+  // map of targets with target name as key
```
Comment (grafnu) [inline proto/forch_configuration.proto:182]: "target with target name" is a bit of a circular definition. It's a path prefix as key, right?

File: bin/stack_functions (hunk 12 above)
Comment (anurag6) [reply bin/stack_functions]: Addressed.

File: bin/test_stack (hunk 13 above)
Comment (anurag6) [reply bin/test_stack]: Addressed.

File: proto/forch_configuration.proto (hunk 14 above)
Comment (anurag6) [reply proto/forch_configuration.proto]: Addressed.

File: proto/forch_configuration.proto (hunk 15 above)
Comment (anurag6) [reply proto/forch_configuration.proto:182]: Addressed.

**Comment by anurag6** [comment]
Addressed concerns. @grafnu  PTAL

File: .travis.yml [hunk 16]
```diff
@@ -9,13 +9,11 @@ addons:
   apt:
     update: true
 install:
-  - set -e
```
Comment (grafnu) [inline .travis.yml:12]: restore

File: bin/stack_functions [hunk 17]
```diff
@@ -400,11 +400,13 @@ function validate_radius_varz {
 
 function test_varz_value {
     varz=$1
-    if [ -z "$2" ]; then
-        port_path=8080/$2
+    port=$2
+    path=$3
+    if [ -z "$3" ]; then
```
Comment (grafnu) [inline bin/stack_functions:405]: use $path here

File: bin/stack_functions [hunk 18]
```diff
@@ -400,11 +400,13 @@ function validate_radius_varz {
...
+        query_path=0.0.0.0:$port
     else
-        port_path=8302
+        query_path=0.0.0.0:$port/$path
```
Comment (grafnu) [inline bin/stack_functions:408]: If statement is likely unnecessary, since 0.0.0.0:8080 is equivalent to 0.0.0.0:8080/

File: bin/test_stack [hunk 19]
```diff
@@ -79,10 +79,11 @@ sudo ip link set t1sw2-eth28 up
...
 test_forch -healthy
-test_varz_value process_state{process=\"faucet\"} forch
-test_varz_value process_state{process=\"gauge\"} forch
-test_varz_value process_state{process=\"forch\"} forch
-test_varz_value process_state{process=\"sleep\"} forch
+test_varz_value process_state{process=\"faucet\"} 8302
+test_varz_value process_state{process=\"gauge\"} 8080 forch
+test_varz_value process_state{process=\"forch\"} 8080 forch
+test_varz_value process_state{process=\"sleep\"} 8080 forch
+test_varz_value process_state{process=\"sleep\"} 8080 random_string
```
Comment (grafnu) [inline bin/test_stack:86]: https://xkcd.com/221/ -- not really random :-) -- better use something like unknown or error -- the intention behind it

File: .travis.yml (hunk 16 above)
Comment (anurag6) [reply .travis.yml]: Addressed.

File: bin/stack_functions (hunk 17 above)
Comment (anurag6) [reply bin/stack_functions]: Obsolete.

File: bin/stack_functions (hunk 18 above)
Comment (anurag6) [reply bin/stack_functions]: Addressed.

File: bin/test_stack (hunk 19 above)
Comment (anurag6) [reply bin/test_stack]: Addressed

File: bin/test_stack [hunk 20]
```diff
@@ -83,7 +83,7 @@ test_varz_value process_state{process=\"faucet\"} 8302
 test_varz_value process_state{process=\"gauge\"} 8080 forch
 test_varz_value process_state{process=\"forch\"} 8080 forch
 test_varz_value process_state{process=\"sleep\"} 8080 forch
-test_varz_value process_state{process=\"sleep\"} 8080 random_string
+test_varz_value process_state{process=\"sleep\"} 8080 jkdhfgjksdhg
```
Comment (grafnu) [inline bin/test_stack:86]: Now that's what I'm talkin' about!

**Pull request merged**